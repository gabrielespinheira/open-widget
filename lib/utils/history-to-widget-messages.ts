import { MessageType } from "@lib/types";
import { ChatHistoryMessageType } from "@lib/types/schemas.backend";
import { genId } from "./genId";

function historyToWidgetMessages(mgs: ChatHistoryMessageType[]) {
    const messages: MessageType[] = [];
    for (let i = 0; i < mgs.length; i++) {
        const msg = mgs[i];
        if (msg.from_user === true) {
            if (msg.message && msg.message.length > 0) {
                messages.push({
                    type: "FROM_USER",
                    content: msg.message,
                    id: msg.id.toString(),
                    session_id: msg.session_id ?? "",
                    timestamp: msg.created_at ?? "",
                    serverId: msg.id.toString(),
                });
            }
        }
        else {
            switch (msg.type) {
                case "message":
                    messages.push({
                        type: "FROM_BOT",
                        component: "TEXT",
                        data: {
                            message: msg.message ?? "",
                        },
                        id: msg.id.toString() ?? genId(),
                        serverId: msg.id ?? genId(),
                        timestamp: msg.created_at ?? "",
                        original: msg,
                    });
                    break;
                case "agent_message":
                    messages.push({
                        type: "FROM_BOT",
                        component: "TEXT",
                        data: {
                            message: msg.message ?? "",
                        },
                        id: msg.id.toString() ?? genId(),
                        serverId: msg.id ?? genId(),
                        timestamp: msg.created_at ?? "",
                        original: msg,
                    });
                    break;
                default:
                    messages.push({
                        type: "FROM_BOT",
                        component: "CHAT_EVENT",
                        data: {
                            event: msg.type,
                            message: msg.message
                        },
                        id: msg.id.toString() ?? genId(),
                        serverId: msg.id ?? genId(),
                        original: msg,
                        timestamp: msg.created_at ?? "",
                    });
            }
        }
    }
    return messages;
}

export { historyToWidgetMessages };