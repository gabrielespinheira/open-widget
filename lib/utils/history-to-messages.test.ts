import { ChatHistoryMessageType } from "@lib/types/schemas.backend";
import { describe, expect, it } from "vitest";
import { historyToWidgetMessages } from "./history-to-widget-messages";

const sampleData: ChatHistoryMessageType[] = [
    {
        id: 164240,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: true,
        message: "I need help",
        created_at: "2024-09-30T17:28:21.000Z",
        updated_at: "2024-09-30T17:28:21.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "message",
        agent_id: null,
        agent_name: null,
        agent_avatar: "",
        handoff_happened_during_office_hours: false,
    },
    {
        id: 164241,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: false,
        message:
            "Hi there! How can I assist you today with your Mollie account or services? Let me know what you need help with, and I'll do my best to assist you.",
        created_at: "2024-09-30T17:28:25.000Z",
        updated_at: null,
        debug_json: { calledActions: [], actionSearchQuery: "help", knowledge: [] },
        api_called: false,
        knowledgebase_called: true,
        extra_params: {},
        type: "message",
        agent_id: null,
        agent_name: null,
        agent_avatar: null,
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165176,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: true,
        message: "hi there ",
        created_at: "2024-10-05T23:51:59.000Z",
        updated_at: "2024-10-05T23:51:59.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "message",
        agent_id: null,
        agent_name: null,
        agent_avatar: "",
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165177,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: false,
        message:
            "Hi! How can I assist you today with your Mollie account or services? If you have any questions or need help with something specific, just let me know!",
        created_at: "2024-10-05T23:52:03.000Z",
        updated_at: null,
        debug_json: {
            actionCalls: [],
            actionSearchQuery: "help, Mollie account, services, assist",
            knowledge: [],
        },
        api_called: false,
        knowledgebase_called: true,
        extra_params: {},
        type: "message",
        agent_id: null,
        agent_name: null,
        agent_avatar: null,
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165178,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: false,
        message: "Agent Ahmad Falta took over the chat from the AI agent",
        created_at: "2024-10-05T23:52:35.000Z",
        updated_at: "2024-10-05T23:52:35.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "agent_took_session_from_ai",
        agent_id: 9000002,
        agent_name: "Ahmad Falta",
        agent_avatar: null,
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165179,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: false,
        message: "help me please!",
        created_at: "2024-10-05T23:52:35.000Z",
        updated_at: "2024-10-05T23:52:35.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "agent_message",
        agent_id: 9000002,
        agent_name: "Ahmad Falta",
        agent_avatar: null,
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165180,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: false,
        message: "i am trapped in here ",
        created_at: "2024-10-05T23:52:42.000Z",
        updated_at: "2024-10-05T23:52:42.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "agent_message",
        agent_id: 9000002,
        agent_name: "Ahmad Falta",
        agent_avatar: null,
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165181,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: true,
        message: "how can i help you ? ",
        created_at: "2024-10-05T23:53:05.000Z",
        updated_at: "2024-10-05T23:53:05.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "message",
        agent_id: null,
        agent_name: null,
        agent_avatar: "",
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165182,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: false,
        message: "helpoo ",
        created_at: "2024-10-05T23:54:34.000Z",
        updated_at: "2024-10-05T23:54:34.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "agent_message",
        agent_id: 9000002,
        agent_name: "Ahmad Falta",
        agent_avatar: null,
        handoff_happened_during_office_hours: false,
    },
    {
        id: 165183,
        chatbot_id: "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
        session_id: "fe726b6f-e668-441a-bed6-94f770574953",
        from_user: false,
        message: "will u help me ? ",
        created_at: "2024-10-05T23:55:08.000Z",
        updated_at: "2024-10-05T23:55:08.000Z",
        debug_json: null,
        api_called: false,
        knowledgebase_called: false,
        extra_params: null,
        type: "agent_message",
        agent_id: 9000002,
        agent_name: "Ahmad Falta",
        agent_avatar: null,
        handoff_happened_during_office_hours: false,
    },
];


describe("history-to-messages", () => {
    it("should convert history to messages", () => {
        const output = historyToWidgetMessages(sampleData);
        expect(output).toMatchSnapshot();
    });
});