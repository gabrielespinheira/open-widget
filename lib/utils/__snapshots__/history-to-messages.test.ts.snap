// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`history-to-messages > should convert history to messages 1`] = `
[
  {
    "content": "I need help",
    "deliveredAt": null,
    "id": "164240",
    "serverId": "164240",
    "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
    "timestamp": "2024-09-30T17:28:21.000Z",
    "type": "FROM_USER",
  },
  {
    "agent": {
      "is_ai": true,
      "name": "",
    },
    "component": "TEXT",
    "data": {
      "message": "Hi there! How can I assist you today with your Mollie account or services? Let me know what you need help with, and I'll do my best to assist you.",
    },
    "id": "164241",
    "original": {
      "agent_avatar": null,
      "agent_id": null,
      "agent_name": null,
      "api_called": false,
      "chatbot_id": "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
      "created_at": "2024-09-30T17:28:25.000Z",
      "debug_json": {
        "actionSearchQuery": "help",
        "calledActions": [],
        "knowledge": [],
      },
      "extra_params": {},
      "from_user": false,
      "handoff_happened_during_office_hours": false,
      "id": 164241,
      "knowledgebase_called": true,
      "message": "Hi there! How can I assist you today with your Mollie account or services? Let me know what you need help with, and I'll do my best to assist you.",
      "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
      "type": "message",
      "updated_at": null,
    },
    "serverId": 164241,
    "timestamp": "2024-09-30T17:28:25.000Z",
    "type": "FROM_BOT",
  },
  {
    "content": "hi there ",
    "deliveredAt": null,
    "id": "165176",
    "serverId": "165176",
    "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
    "timestamp": "2024-10-05T23:51:59.000Z",
    "type": "FROM_USER",
  },
  {
    "agent": {
      "is_ai": true,
      "name": "",
    },
    "component": "TEXT",
    "data": {
      "message": "Hi! How can I assist you today with your Mollie account or services? If you have any questions or need help with something specific, just let me know!",
    },
    "id": "165177",
    "original": {
      "agent_avatar": null,
      "agent_id": null,
      "agent_name": null,
      "api_called": false,
      "chatbot_id": "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
      "created_at": "2024-10-05T23:52:03.000Z",
      "debug_json": {
        "actionCalls": [],
        "actionSearchQuery": "help, Mollie account, services, assist",
        "knowledge": [],
      },
      "extra_params": {},
      "from_user": false,
      "handoff_happened_during_office_hours": false,
      "id": 165177,
      "knowledgebase_called": true,
      "message": "Hi! How can I assist you today with your Mollie account or services? If you have any questions or need help with something specific, just let me know!",
      "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
      "type": "message",
      "updated_at": null,
    },
    "serverId": 165177,
    "timestamp": "2024-10-05T23:52:03.000Z",
    "type": "FROM_BOT",
  },
  {
    "agent": {
      "is_ai": true,
      "name": "Ahmad Falta",
    },
    "component": "CHAT_EVENT",
    "data": {
      "event": "agent_took_session_from_ai",
      "message": "Agent Ahmad Falta took over the chat from the AI agent",
    },
    "id": "165178",
    "original": {
      "agent_avatar": null,
      "agent_id": 9000002,
      "agent_name": "Ahmad Falta",
      "api_called": false,
      "chatbot_id": "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
      "created_at": "2024-10-05T23:52:35.000Z",
      "debug_json": null,
      "extra_params": null,
      "from_user": false,
      "handoff_happened_during_office_hours": false,
      "id": 165178,
      "knowledgebase_called": false,
      "message": "Agent Ahmad Falta took over the chat from the AI agent",
      "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
      "type": "agent_took_session_from_ai",
      "updated_at": "2024-10-05T23:52:35.000Z",
    },
    "serverId": 165178,
    "timestamp": "2024-10-05T23:52:35.000Z",
    "type": "FROM_BOT",
  },
  {
    "agent": {
      "agent_avatar": "",
      "is_ai": false,
      "name": "Ahmad Falta",
    },
    "component": "TEXT",
    "data": {
      "message": "help me please!",
    },
    "id": "165179",
    "original": {
      "agent_avatar": null,
      "agent_id": 9000002,
      "agent_name": "Ahmad Falta",
      "api_called": false,
      "chatbot_id": "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
      "created_at": "2024-10-05T23:52:35.000Z",
      "debug_json": null,
      "extra_params": null,
      "from_user": false,
      "handoff_happened_during_office_hours": false,
      "id": 165179,
      "knowledgebase_called": false,
      "message": "help me please!",
      "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
      "type": "agent_message",
      "updated_at": "2024-10-05T23:52:35.000Z",
    },
    "serverId": 165179,
    "timestamp": "2024-10-05T23:52:35.000Z",
    "type": "FROM_BOT",
  },
  {
    "agent": {
      "agent_avatar": "",
      "is_ai": false,
      "name": "Ahmad Falta",
    },
    "component": "TEXT",
    "data": {
      "message": "i am trapped in here ",
    },
    "id": "165180",
    "original": {
      "agent_avatar": null,
      "agent_id": 9000002,
      "agent_name": "Ahmad Falta",
      "api_called": false,
      "chatbot_id": "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
      "created_at": "2024-10-05T23:52:42.000Z",
      "debug_json": null,
      "extra_params": null,
      "from_user": false,
      "handoff_happened_during_office_hours": false,
      "id": 165180,
      "knowledgebase_called": false,
      "message": "i am trapped in here ",
      "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
      "type": "agent_message",
      "updated_at": "2024-10-05T23:52:42.000Z",
    },
    "serverId": 165180,
    "timestamp": "2024-10-05T23:52:42.000Z",
    "type": "FROM_BOT",
  },
  {
    "content": "how can i help you ? ",
    "deliveredAt": null,
    "id": "165181",
    "serverId": "165181",
    "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
    "timestamp": "2024-10-05T23:53:05.000Z",
    "type": "FROM_USER",
  },
  {
    "agent": {
      "agent_avatar": "",
      "is_ai": false,
      "name": "Ahmad Falta",
    },
    "component": "TEXT",
    "data": {
      "message": "helpoo ",
    },
    "id": "165182",
    "original": {
      "agent_avatar": null,
      "agent_id": 9000002,
      "agent_name": "Ahmad Falta",
      "api_called": false,
      "chatbot_id": "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
      "created_at": "2024-10-05T23:54:34.000Z",
      "debug_json": null,
      "extra_params": null,
      "from_user": false,
      "handoff_happened_during_office_hours": false,
      "id": 165182,
      "knowledgebase_called": false,
      "message": "helpoo ",
      "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
      "type": "agent_message",
      "updated_at": "2024-10-05T23:54:34.000Z",
    },
    "serverId": 165182,
    "timestamp": "2024-10-05T23:54:34.000Z",
    "type": "FROM_BOT",
  },
  {
    "agent": {
      "agent_avatar": "",
      "is_ai": false,
      "name": "Ahmad Falta",
    },
    "component": "TEXT",
    "data": {
      "message": "will u help me ? ",
    },
    "id": "165183",
    "original": {
      "agent_avatar": null,
      "agent_id": 9000002,
      "agent_name": "Ahmad Falta",
      "api_called": false,
      "chatbot_id": "23ab3726-187f-4c11-83e9-0fb3ccabd6f4",
      "created_at": "2024-10-05T23:55:08.000Z",
      "debug_json": null,
      "extra_params": null,
      "from_user": false,
      "handoff_happened_during_office_hours": false,
      "id": 165183,
      "knowledgebase_called": false,
      "message": "will u help me ? ",
      "session_id": "fe726b6f-e668-441a-bed6-94f770574953",
      "type": "agent_message",
      "updated_at": "2024-10-05T23:55:08.000Z",
    },
    "serverId": 165183,
    "timestamp": "2024-10-05T23:55:08.000Z",
    "type": "FROM_BOT",
  },
]
`;
